### PR TITLE
Status Monitor can monitor Multiple Servers

### PR DESCRIPTION
**Pull request by mrtracy**
Status monitor tool now accepts a comma-delimited list of addresses on the
command line, and will monitor all given servers in parallel.

The output log has been removed; it is not being used and is not particularly
parseable at the current time.

The next step is to modify the cockroach server to expose a more general "/Health"
endpoint, which will indicate only if the individual node is running.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3168)

### CODE REVIEW

**Comment by BramGruneir** [comment]
I don't know if a lot of these changes are necessary. 
The status server already internally can redirect to other nodes and this way, you can have the monitor hit the load balancer instead of having to pass in a list of addrs.  
If you get the node statuses, you can parse the list of node ids (say once every min or so), and then send all your individual node health requests through any node.  This will allow the monitor to work if new nodes are added and removed without having to restart it.

Also, there already exists a health of sorts (which really should be augmented) 
```
	// statusDetailsPattern exposes a node's details.
	statusDetailsPattern = "/_status/details/:node_id"
```

File: cmd/monitor/main.go [hunk 1]
```diff
@@ -36,15 +38,19 @@ import (
...
-var outputDir = flag.String("dir", ".", "Directory in which to output the status logs to.")
-var interval = flag.Duration("interval", 10*time.Second, "Interval in which to poll the cluster's status.")
-var addr = flag.String("addr", ":26257", "The host:port of the cockroach cluster.")
+const (
+	// urlPath is the http path of the status server on each node.
+	urlPath = "_status/nodes"
+)
+
+var interval = flag.Duration("interval", 10*time.Second, "Interval in which to poll each services's status.")
+var addrs = flag.String("addrs", ":26257", "Comma-separated list of host:port status endpoints to monitor.")
```
Comment (mberhault) [inline cmd/monitor/main.go]: just "host:port to monitor". including "status endpoints" in the message makes it sounds like I should pass host:port/url

**Comment by mberhault** [comment]
@BramGruneir: the goal of this tool is to have it watch a running cluster and notice when individual nodes are down. In that scenario, using the load balancer (or any other round-robin or random thing) will not help as it will only redirect to the healthy instances. We also don't want to rely on a given node knowing the list of other running nodes. Instead, the list of addresses will come from the terraform script which know which ones should be up.

As for the status endpoint, we really need a custom one that is only about health. eg: /healthz or something which returns 200 and response text OK when the server is running. This is pretty standard behavior, and we really should rely on other endpoints that may have more side effects (eg: we can't rely on something that reads TS data, requires gossip to be connected, etc...)

**Comment by BramGruneir** [comment]
@mberhault  

What's wrong with using gossip to get the address in the network instead of terraform?  If any node doesn't know about another node, it probably means that it's down or not connected to the cluster.
Any examples that kill and restart nodes are going to have to have specific logic written into them if that's the case.  
Also, we redirect status requests to the correct node already, but we just do it based on node number instead of address.

Also, a custom endpoint is cool, but it just seemed like extra work at the moment when we have an endpoint that returns basic info from a running node already (and fails if that node is down.)

File: cmd/monitor/main.go (hunk 1 above)
Comment (mrtracy) [inline cmd/monitor/main.go]: Done

**Comment by mrtracy** [comment]
I changed the endpoint to point at `/details/local`, which returns quickly without consulting other nodes.  The tool still monitors a definite list of addresses, but I think we've agreed in discussion and offline that this is best for our current use case.

File: cmd/monitor/main.go [hunk 2]
```diff
@@ -52,71 +58,113 @@ var retryOptions = retry.Options{
...
+		}
+		log.Infof("Monitoring Status URL: %s\n", client.url)
+		stopper.RunWorker(func() {
+			timer := time.Tick(*interval)
+			for {
+				select {
+				case <-stopper.ShouldStop():
+					return
+				case <-timer:
+					_, err := client.queryStatus()
```
Comment (BramGruneir) [inline cmd/monitor/main.go]: if _, err := client.querystatus(); err != nil {

File: cmd/monitor/main.go [hunk 3]
```diff
@@ -52,71 +58,113 @@ var retryOptions = retry.Options{
...
+	var err error
+	monitor.httpClient, err = context.GetHTTPClient()
+	if err != nil {
+		return nil, err
+	}
+	monitor.url = fmt.Sprintf("%s://%s/%s", context.HTTPRequestScheme(), monitor.addr, urlPath)
+	return monitor, nil
+}
+
+func (m *statusMonitor) queryStatus() ([]byte, error) {
```
Comment (BramGruneir) [inline cmd/monitor/main.go]: Does it make sense to return the body if you're not consuming it?

File: cmd/monitor/main.go [hunk 4]
```diff
@@ -52,71 +58,113 @@ var retryOptions = retry.Options{
...
+		stopper.RunWorker(func() {
+			timer := time.Tick(*interval)
+			for {
+				select {
+				case <-stopper.ShouldStop():
+					return
+				case <-timer:
+					_, err := client.queryStatus()
+					if err != nil {
+						log.Warningf("Could not get status from url %s. Time since monitor started %s.", client.url, time.Since(startTime))
```
Comment (BramGruneir) [inline cmd/monitor/main.go]: Perhaps preface all log messages with a "time since monitor started"?

Comment (mrtracy) [inline cmd/monitor/main.go]: Done

File: cmd/monitor/main.go (hunk 3 above)
Comment (mrtracy) [inline cmd/monitor/main.go]: Removed

File: cmd/monitor/main.go (hunk 2 above)
Comment (mrtracy) [inline cmd/monitor/main.go]: Done

**Comment by mrtracy** [comment]
ping @BramGruneir @mberhault

**Comment by BramGruneir** [comment]
LGTM

**Pull request merged**